File: front/composables/ColorImage.ts
import type { BoundingBox, Position } from "./useImage";

export class ColorImage {
  private name: string;
  private pixels: Uint8ClampedArray;
  private w: number;
  private h: number;
  private cache: Uint8ClampedArray;

  public constructor(width: number, height: number) {
    this.name = "image" + Math.floor(Math.random() * 10);
    this.w = width;
    this.h = height;
    this.pixels = new Uint8ClampedArray(width * height * 4);
    this.cache = new Uint8ClampedArray(width * height * 4);
  }

  public get image(): Uint8ClampedArray {
    return this.cache
  }

  private isDrawing = false
  private previousePos: Position = { x: 0, y: 0, pressure: 0 }

  private plot(x: number, y: number): void {
    const idx = (y * this.w + x) * 4
    this.pixels[idx] = this.pixels[idx + 1] = this.pixels[idx + 2] = 0
    this.pixels[idx + 3] = 255
  }

  private line(pos1: Position, pos2: Position): void {
    const x0 = Math.round(pos1.x)
    const y0 = Math.round(pos1.y)
    const x1 = Math.round(pos2.x)
    const y1 = Math.round(pos2.y)
    const dx = Math.abs(x1 - x0)
    const dy = Math.abs(y1 - y0)
    const sx = x0 < x1 ? 1 : -1
    const sy = y0 < y1 ? 1 : -1
    let err = dx - dy

    let x = x0
    let y = y0
    while (true) {
      this.plot(x, y)
      if (x === x1 && y === y1) break
      const e2 = 2 * err
      if (e2 > -dy) {
        err -= dy
        x += sx
      }
      if (e2 < dx) {
        err += dx
        y += sy
      }
    }
  }

  public stroke(pos: Position): BoundingBox {
    let result: BoundingBox = { left: 0, top: 0, width: 0, height: 0 }
    if (this.isDrawing) {
      if (pos.pressure === 0) {
        this.isDrawing = false

        // undoバッファに突っ込む。
        // undoBuffer.push(new ImageData(new Uint8ClampedArray(pixel), width));
      }
      this.line(this.previousePos, pos)
      result = {
        left: Math.floor(Math.min(pos.x, this.previousePos.x)),
        top: Math.floor(Math.min(pos.y, this.previousePos.y)),
        width: Math.ceil(Math.abs(this.previousePos.x - pos.x)) + 2,
        height: Math.ceil(Math.abs(this.previousePos.y - pos.y)) + 2,
      }
      this.previousePos = pos
    } else {
      this.isDrawing = true
      result = {
        left: Math.floor(pos.x),
        top: Math.floor(pos.y),
        width: 1,
        height: 1,
      }
      this.previousePos = pos
      this.plot(Math.round(pos.x), Math.round(pos.y))
    }
    this.cache.set(this.pixels)
    return result
  }
}

File: wasm/src/layer/color_image.rs
use super::{BoundingBox, ImageLayer};
use crate::pen::DrawingPen;
use crate::utils;
use std::convert::TryInto;
extern crate web_sys;

#[allow(unused_macros)]
macro_rules! log {
    ( $( $t:tt )* ) => {
        web_sys::console::log_1(&format!( $( $t )* ).into());
    }
}

// #[derive(Debug)]
pub struct ColorImage {
    pixels: Vec<u8>,
    width: i32,
    height: i32,
    is_drawing: bool,
    previouse_pos: super::Position,
    prev_pxs: Vec<u8>,
}

impl ImageLayer for ColorImage {
    fn stroke(&mut self, x: f64, y: f64, pressure: f64, pen: &Box<dyn DrawingPen>) -> BoundingBox {
        let radius = pen.size().ceil() as i32;
        let mut result = BoundingBox {
            left: 0,
            top: 0,
            width: 0,
            height: 0,
        };
        if self.is_drawing {
            if pressure == 0.0 {
                self.is_drawing = false;
                self.prev_pxs = self.pixels.clone();

                // undoバッファに突っ込む。
            };
            self.line(
                &super::Position {
                    x: x,
                    y: y,
                    pressure: pressure,
                },
                pen,
            );
            result = BoundingBox {
                left: x.min(self.previouse_pos.x).floor() as i32 - radius,
                top: y.min(self.previouse_pos.y).floor() as i32 - radius,
                width: (self.previouse_pos.x - x).abs().ceil() as i32 + 1 + radius + radius,
                height: (self.previouse_pos.y - y).abs().ceil() as i32 + 1 + radius + radius,
            };
            self.previouse_pos = super::Position {
                x: x,
                y: y,
                pressure: pressure,
            };
        } else {
            self.is_drawing = true;
            result = BoundingBox {
                left: x.floor() as i32 - radius,
                top: y.floor() as i32 - radius,
                width: 1 + radius + radius,
                height: 1 + radius + radius,
            };
            self.previouse_pos = super::Position {
                x: x,
                y: y,
                pressure: pressure,
            };
            self.line(
                &super::Position {
                    x: x,
                    y: y,
                    pressure: pressure,
                },
                pen,
            )
        }

        result
    }

    fn pixel_pointer(&self) -> *const u8 {
        self.pixels.as_ptr()
    }

    fn pixel(&mut self) -> &mut Vec<u8> {
        &mut self.pixels
    }
}

impl ColorImage {
    pub fn new(width: i32, height: i32) -> ColorImage {
        utils::set_panic_hook();
        // Todo: widthとheightの範囲をチェックしないといけないけどあとで。
        let length: usize = (width as u32 * height as u32 * 4).try_into().unwrap();
        ColorImage {
            pixels: vec![0; length],
            width: width,
            height: height,
            is_drawing: false,
            previouse_pos: super::Position {
                x: 0.0,
                y: 0.0,
                pressure: 0.0,
            },
            prev_pxs: vec![0; length],
        }
    }

    /// p1からp2まで直線を描画する。
    fn line(&mut self, p2: &super::Position, pen: &Box<dyn DrawingPen>) {
        pen.line(
            &self.previouse_pos,
            p2,
            self.width,
            self.height,
            &mut self.pixels,
        )
    }
}
